implementation/src/test/java/io/smallrye/mutiny/operators/UniCacheTest.java

```diff
@@ -273,4 +366,154 @@ public void onFailure(Throwable ignored) {
+        Uni uni = Uni.createFrom().deferred(() -> Uni.createFrom().item(subscriptionCount.getAndIncrement()))
+                .cacheEvents().atMost(Duration.ofMillis(100));
+
+        UniAssertSubscriber<Integer> sub1 = new UniAssertSubscriber<>();
+        UniAssertSubscriber<Integer> sub2 = new UniAssertSubscriber<>();
+        UniAssertSubscriber<Integer> sub3 = new UniAssertSubscriber<>();
+        UniAssertSubscriber<Integer> sub4 = new UniAssertSubscriber<>();
+
+        uni.subscribe().withSubscriber(sub1);
+        uni.subscribe().withSubscriber(sub2);
+
+        Uni.createFrom().item(0).onItem().delayIt().by(Duration.ofMillis(100)).await().indefinitely();
+
+        uni.subscribe().withSubscriber(sub3);
+        uni.subscribe().withSubscriber(sub4);
+
+        assertThat(subscriptionCount.get()).isEqualTo(2);
+        sub1.assertItem(0);
+        sub2.assertItem(0);
+        sub3.assertItem(1);
+        sub3.assertItem(1);
+    }
+
+    @Test
+    public void testThatFailureIsRenewedEvenBeforeCacheDurationExceeded() {
+        AtomicInteger subscriptionCount = new AtomicInteger(0);
+        Uni uni = Uni.createFrom()
+                .deferred(() -> Uni.createFrom().failure(new Exception("" + subscriptionCount.getAndIncrement())))
+                .cacheEvents().atMost(Duration.ofMinutes(1));
+
+        UniAssertSubscriber<Integer> sub1 = new UniAssertSubscriber<>();
+        UniAssertSubscriber<Integer> sub2 = new UniAssertSubscriber<>();
+        UniAssertSubscriber<Integer> sub3 = new UniAssertSubscriber<>();
+        UniAssertSubscriber<Integer> sub4 = new UniAssertSubscriber<>();
+
+        uni.subscribe().withSubscriber(sub1);
+        uni.subscribe().withSubscriber(sub2);
+        uni.subscribe().withSubscriber(sub3);
+        uni.subscribe().withSubscriber(sub4);
+
+        assertThat(subscriptionCount.get()).isEqualTo(4);
+        sub1.assertFailure(Exception.class, "0");
+        sub2.assertFailure(Exception.class, "1");
+        sub3.assertFailure(Exception.class, "2");
+        sub4.assertFailure(Exception.class, "3");
+    }
+
+    @Test
+    public void testThatOnlyInvalidCacheEventsDoNotRace() {
+        AtomicInteger cacheValidationCount = new AtomicInteger(0);
+        // predicate counts validations but is always `false` (event always invalid, every test causes resubscription)
+        BiPredicate<Integer, Throwable> validator = (i, f) -> cacheValidationCount.incrementAndGet() == 0;
+        AtomicInteger sourceSubscriptionCount = new AtomicInteger(0);
+        Uni<Integer> uni = Uni.createFrom().deferred(() -> Uni.createFrom().item(sourceSubscriptionCount.getAndIncrement()))
+                .cacheEvents().whilst(validator);
+
+        // Every second subscription causes a re-subscription to the upstream
+        // This must not cause race conditions
+        List<UniAssertSubscriber<Integer>> results = Collections.synchronizedList(new LinkedList<>());
+        Executor parallel = Executors.newFixedThreadPool(1);
```

Comment:
This test floods the UniCache with subscriptions while the cached item never becomes valid.
I'm not getting this to succeed when running parallel executors (threadpool > 1).
Also have currently no idea where exactly the problem lies, fails because not all Subscribers got a item or are triggered at all...